# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/indexedDB/ActorsParent.cpp
# Commit: fcea7ad7ad7b
# Full Hash: fcea7ad7ad7b6055094fc567b5716d27d55c55fc
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2019-12-05 04:11:08
# Regressor Bug: 1600906
# File Overlap Count: 1
# Description:
#   Bug 1600906 - Make use of FlippedOnce in FullIndexMetadata. r=dom-workers-and-storage-reviewers,ytausky
#   
#   Depends on D55675
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D55676
# ==============================================================================

diff -r 69dba2abeaf4 -r fcea7ad7ad7b dom/indexedDB/ActorsParent.cpp
--- a/dom/indexedDB/ActorsParent.cpp	Wed Dec 04 14:31:13 2019 +0000
+++ b/dom/indexedDB/ActorsParent.cpp	Wed Dec 04 14:30:55 2019 +0000
@@ -346,20 +346,14 @@
  * Metadata classes
  ******************************************************************************/
 
+// Can be instantiated either on the QuotaManager IO thread or on a
+// versionchange transaction thread. These threads can never race so this is
+// totally safe.
 struct FullIndexMetadata {
-  IndexMetadata mCommonMetadata;
-
-  bool mDeleted;
-
- public:
-  FullIndexMetadata()
-      : mCommonMetadata(0, nsString(), KeyPath(0), nsCString(), false, false,
-                        false),
-        mDeleted(false) {
-    // This can happen either on the QuotaManager IO thread or on a
-    // versionchange transaction thread. These threads can never race so this is
-    // totally safe.
-  }
+  IndexMetadata mCommonMetadata = {0,     nsString(), KeyPath(0), nsCString(),
+                                   false, false,      false};
+
+  FlippedOnce<false> mDeleted;
 
   NS_INLINE_DECL_THREADSAFE_REFCOUNTING(FullIndexMetadata)
 
@@ -15113,13 +15107,13 @@
     return IPC_FAIL_NO_REASON(this);
   }
 
-  foundIndexMetadata->mDeleted = true;
+  foundIndexMetadata->mDeleted.Flip();
 
   DebugOnly<bool> foundTargetId = false;
   const bool isLastIndex =
       std::all_of(foundObjectStoreMetadata->mIndexes.cbegin(),
                   foundObjectStoreMetadata->mIndexes.cend(),
-                  [&foundTargetId, aIndexId](const auto& indexEntry) {
+                  [&foundTargetId, aIndexId](const auto& indexEntry) -> bool {
                     if (uint64_t(aIndexId) == indexEntry.GetKey()) {
                       foundTargetId = true;
                       return true;
